### PR TITLE
Handle nested records in cut

### PR DESCRIPTION
**Pull request by aswan**
This got a lot more gnarly that I anticipated but even if they're infrequently used, expressions that combine different fields like `cut a, b.c, b.d, x.y.z` are tricky to handle correctly and efficiently.

### CODE REVIEW

**Comment by aswan** [comment]
note to reviewers: I tried to break this into some smaller pieces so the commits can be reviewed individually.  In particular, the first is pretty boilerplate, the second is where all the real magic happens, and the third is new test cases for the newly added code.

**Comment by henridf** [comment]
Just started off experimenting with this before reviewing the change. 
Looks good, but I did run in to this corner case:

```
$ cat conn.log | zq  "cut id, id.orig_h | tail 1" -
#0:record[id:record[orig_h:addr,orig_p:port,resp_h:addr,resp_p:port],id:record[orig_h:addr]]
```

I'm not sure if this should be a cut error or if the second cut field `id.orig_h` should be silently "swallowed" by the first one such that `cut r, r.x` would be give identical results to `cut r`.

(And out of scope of this PR, but it shouldn't be possible to have same-named fields in a record...)

**Comment by aswan** [comment]
> Just started off experimenting with this before reviewing the change.
> Looks good, but I did run in to this corner case:
> 
> ```
> $ cat conn.log | zq  "cut id, id.orig_h | tail 1" -
> #0:record[id:record[orig_h:addr,orig_p:port,resp_h:addr,resp_p:port],id:record[orig_h:addr]]
> ```
> 
> I'm not sure if this should be a cut error or if the second cut field `id.orig_h` should be silently "swallowed" by the first one such that `cut r, r.x` would be give identical results to `cut r`.
> 
> (And out of scope of this PR, but it shouldn't be possible to have same-named fields in a record...)

Ah good catch, a simpler case is that eg `cut uid, uid` works today, this is a more subtle manifestation of that issue.  I guess we have the same issue in groupby (e.g., `something() by uid, uid`)

If its okay with you, I'll file a follow-up issue for that.  I don't think it will be difficult to fix but want to avoid getting too much stuff into in already large PR.

**Comment by henridf** [comment]
@aswan ah i hadn't even noticed the generality. yes agreed that solving it doesn't have to be done here. reviewing now...

File: zql/zql.peg [hunk 1]
```diff
@@ -260,6 +260,20 @@ fieldExprList
       RETURN(result)
   }
 
+fieldRefDotOnly
```
Comment (henridf) [inline zql/zql.peg]: Seems like there's some duplication with fieldReference but we can tidy this up later...

File: proc/cut.go [hunk 2]
```diff
@@ -5,74 +5,268 @@ import (
...
 		}
-		columns = append(columns, in.Type.Columns[colno])
 	}
-	out := c.Resolver.GetByColumns(columns)
-	c.cutmap[in.ID] = out
-	return out
+	return true
 }
 
 // CreateCut returns a new record value derived by keeping only the fields
```
Comment (henridf) [inline proc/cut.go]: since you're in here, s/CreateCut/cut/

File: proc/cut.go [hunk 3]
```diff
@@ -5,74 +5,268 @@ import (
...
+}
+
+// Split an ast.FieldExpr representing a chain of record field references
+// into a list of strings representing the names.
+// E.g., "x.y.z" -> ["x", "y", "z"]
+func split(node ast.FieldExpr) ([]string, error) {
+	switch n := node.(type) {
+	case *ast.FieldRead:
+		return []string{n.Field}, nil
+	case *ast.FieldCall:
```
Comment (henridf) [inline proc/cut.go]: this is made impossible by the current grammar right? if so i would suggest relaxing the grammar, as it's more helpful to get the error message below than "parse error: 1:1 (0): no match found".

File: proc/cut_test.go [hunk 4]
```diff
@@ -3,76 +3,96 @@ package proc_test
...
-	require.NoError(t, pt.Finish())
+	proc.TestOneProc(t, fooAndBar, fooAndBar, "cut foo,bar")
+}
+
+// Test that illegal cut operations fail at compile time with a
+// reasonable error message.
+func testNonAdjacentFields(t *testing.T, zql string) {
+	_, err := proc.CompileTestProc(zql, nil, nil)
+	require.Error(t, err, "cut with non-adjacent records failed")
+	require.Regexp(t, "All cut fields in record \\S+ must be adjacent", err.Error(), "error message for cutting non-adjacent fields is correct")
```
Comment (henridf) [inline proc/cut_test.go]: Relying on the error text this way is a little fragile... I realize you couldn't just define a static error value and compare to that since you have the record name in the error message (👍 for that). However using the new fancy [wrapped errors](https://blog.golang.org/go1.13-errors), it should be possible to compare to an underlying static error  that is wrapped with the more informative message.

File: proc/cut.go (hunk 2 above)
Comment (aswan) [reply proc/cut.go]: Will do.  I also notice that this function never returns a non-nil error so I'm going to remove it from the signature.

File: proc/cut.go (hunk 3 above)
Comment (henridf) [reply proc/cut.go]: (where by "current" i really meant "new as of this PR"....)

Comment (aswan) [reply proc/cut.go]: That's how I wrote it at first, but the javascript parser gives a better error so in this way, the desktop app catches an invalid proc (e.g., `cut foo[2]`) earlier and with a reasonable error.  I don't feel strongly about this though, happy to change it if you prefer.

Comment (henridf) [reply proc/cut.go]: good points... i feel less strongly now i understand this context, so feel free not to change it...

**Comment by aswan** [comment]
> This looks great. I would create an abstraction for the ColumnBuilder that contains your fieldInfo stuff to represent the column structure then has a method to compute the structured zval from the input zvals. This will be needed by group-by. Maybe best to merge this then when you get to group-by you can make sure this new abstraction works in both places.

Yep, that's precisely what I've been working on...

File: proc/cut_test.go (hunk 4 above)
Comment (aswan) [reply proc/cut_test.go]: I did this by just creating a new struct that implements the error interface, no wrapping involved.  I'm not sure what's idiomatic go, certainly open to further refinements here...
